File: lib/es6/js_list.js


import * as Curry from "./curry.js";
import * as Js_vector from "./js_vector.js";
import * as Caml_option from "./caml_option.js";

function length(l) {
  let _len = 0;
  let _x = l;
  while(true) {
    let x = _x;
    let len = _len;
    if (!x) {
      return len;
    }
    _x = x.tl;
    _len = len + 1 | 0;
    continue;
  };
}

function cons(x, xs) {
  return {
    hd: x,
    tl: xs
  };
}

function isEmpty(x) {
  return x === /* [] */0;
}

function hd(x) {
  if (x) {
    return Caml_option.some(x.hd);
  }
  
}

function tl(x) {
  if (x) {
    return x.tl;
  }
  
}

function nth(l, n) {
  if (n < 0) {
    return;
  }
  let _l = l;
  let _n = n;
  while(true) {
    let n$1 = _n;
    let l$1 = _l;
    if (!l$1) {
      return;
    }
    if (n$1 === 0) {
      return Caml_option.some(l$1.hd);
    }
    _n = n$1 - 1 | 0;
    _l = l$1.tl;
    continue;
  };
}

function revAppend(_l1, _l2) {
  while(true) {
    let l2 = _l2;
    let l1 = _l1;
    if (!l1) {
      return l2;
    }
    _l2 = {
      hd: l1.hd,
      tl: l2
    };
    _l1 = l1.tl;
    continue;
  };
}

function rev(l) {
  return revAppend(l, /* [] */0);
}

function mapRevAux(f, _acc, _ls) {
  while(true) {
    let ls = _ls;
    let acc = _acc;
    if (!ls) {
      return acc;
    }
    _ls = ls.tl;
    _acc = {
      hd: Curry._1(f, ls.hd),
      tl: acc
    };
    continue;
  };
}

function mapRev(f, ls) {
  return mapRevAux(f, /* [] */0, ls);
}

function map(f, ls) {
  return revAppend(mapRevAux(f, /* [] */0, ls), /* [] */0);
}

function iter(f, _x) {
  while(true) {
    let x = _x;
    if (!x) {
      return;
    }
    Curry._1(f, x.hd);
    _x = x.tl;
    continue;
  };
}

function iteri(f, l) {
  let _i = 0;
  let _x = l;
  while(true) {
    let x = _x;
    let i = _i;
    if (!x) {
      return;
    }
    Curry._2(f, i, x.hd);
    _x = x.tl;
    _i = i + 1 | 0;
    continue;
  };
}

function foldLeft(f, _accu, _l) {
  while(true) {
    let l = _l;
    let accu = _accu;
    if (!l) {
      return accu;
    }
    _l = l.tl;
    _accu = Curry._2(f, accu, l.hd);
    continue;
  };
}

function tailLoop(f, _acc, _x) {
  while(true) {
    let x = _x;
    let acc = _acc;
    if (!x) {
      return acc;
    }
    _x = x.tl;
    _acc = Curry._2(f, x.hd, acc);
    continue;
  };
}

function foldRight(f, l, init) {
  let loop = function (n, x) {
    if (!x) {
      return init;
    }
    let t = x.tl;
    let h = x.hd;
    if (n < 1000) {
      return Curry._2(f, h, loop(n + 1 | 0, t));
    } else {
      return Curry._2(f, h, tailLoop(f, init, revAppend(t, /* [] */0)));
    }
  };
  return loop(0, l);
}

function flatten(lx) {
  let _acc = /* [] */0;
  let _lx = lx;
  while(true) {
    let lx$1 = _lx;
    let acc = _acc;
    if (!lx$1) {
      return revAppend(acc, /* [] */0);
    }
    _lx = lx$1.tl;
    _acc = revAppend(lx$1.hd, acc);
    continue;
  };
}

function filterRevAux(f, _acc, _xs) {
  while(true) {
    let xs = _xs;
    let acc = _acc;
    if (!xs) {
      return acc;
    }
    let ys = xs.tl;
    let y = xs.hd;
    if (Curry._1(f, y)) {
      _xs = ys;
      _acc = {
        hd: y,
        tl: acc
      };
      continue;
    }
    _xs = ys;
    continue;
  };
}

function filter(f, xs) {
  return revAppend(filterRevAux(f, /* [] */0, xs), /* [] */0);
}

function filterMapRevAux(f, _acc, _xs) {
  while(true) {
    let xs = _xs;
    let acc = _acc;
    if (!xs) {
      return acc;
    }
    let ys = xs.tl;
    let z = Curry._1(f, xs.hd);
    if (z !== undefined) {
      _xs = ys;
      _acc = {
        hd: Caml_option.valFromOption(z),
        tl: acc
      };
      continue;
    }
    _xs = ys;
    continue;
  };
}

function filterMap(f, xs) {
  return revAppend(filterMapRevAux(f, /* [] */0, xs), /* [] */0);
}

function countBy(f, xs) {
  let _acc = 0;
  let _xs = xs;
  while(true) {
    let xs$1 = _xs;
    let acc = _acc;
    if (!xs$1) {
      return acc;
    }
    _xs = xs$1.tl;
    _acc = Curry._1(f, xs$1.hd) ? acc + 1 | 0 : acc;
    continue;
  };
}

function init(n, f) {
  return Js_vector.toList(Js_vector.init(n, f));
}

function toVector(xs) {
  if (!xs) {
    return [];
  }
  let a = new Array(length(xs));
  let _i = 0;
  let _x = xs;
  while(true) {
    let x = _x;
    let i = _i;
    if (!x) {
      return a;
    }
    a[i] = x.hd;
    _x = x.tl;
    _i = i + 1 | 0;
    continue;
  };
}

function equal(cmp, _xs, _ys) {
  while(true) {
    let ys = _ys;
    let xs = _xs;
    if (!xs) {
      if (ys) {
        return false;
      } else {
        return true;
      }
    }
    if (!ys) {
      return false;
    }
    if (!Curry._2(cmp, xs.hd, ys.hd)) {
      return false;
    }
    _ys = ys.tl;
    _xs = xs.tl;
    continue;
  };
}

export {
  length,
  cons,
  isEmpty,
  hd,
  tl,
  nth,
  revAppend,
  rev,
  mapRev,
  map,
  iter,
  iteri,
  foldLeft,
  foldRight,
  flatten,
  filter,
  filterMap,
  countBy,
  init,
  toVector,
  equal,
}
/* No side effect */

File: jscomp/others/js_list.ml
(* Copyright (C) 2015-2016 Bloomberg Finance L.P.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * In addition to the permissions granted to you by the LGPL, you may combine
 * or link a "work that uses the Library" with a publicly distributed version
 * of this file to produce a combined library or application, then distribute
 * that combined work under the terms of your choosing, with no requirement
 * to comply with the obligations normally placed on you by section 4 of the
 * LGPL version 3 (or the corresponding section of a later version of the LGPL
 * should you choose to use a later version).
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA. *)

(** WIP: List operations *)
type 'a t = 'a list



let rec lengthAux len = function
  | [] -> len
  | _::l -> lengthAux (len + 1) l

let length l = lengthAux 0 l

let cons  x xs = x :: xs

let isEmpty x =  x = []

let hd = function
  | [] -> None
  | a::_ -> Some a

let tl = function
  | [] -> None
  | _::l -> Some l

let nth l n =
  if n < 0 then None else
    let rec nth_aux l n =
      match l with
      | [] -> None
      | a::l -> if n = 0 then Some a else nth_aux l (n-1)
    in nth_aux l n



let rec revAppend l1 l2 =
  match l1 with
  | [] -> l2
  | a :: l -> revAppend l (a :: l2)

let rev l = revAppend l []


let rec mapRevAux f acc ls =
  match ls with
  | [] ->  acc
  | a::l ->
    mapRevAux f (f a [@bs] :: acc) l

let mapRev f ls = mapRevAux f [] ls

let map f ls = rev (mapRevAux f [] ls )

let rec iter f = function
    [] -> ()
  | a::l -> f a [@bs]; iter f l

let rec iteri i f = function
    [] -> ()
  | a::l -> f i a [@bs]; iteri (i + 1) f l

let iteri f l = iteri 0 f l

let rec foldLeft f accu l =
  match l with
    [] -> accu
  | a::l -> foldLeft f (f accu a [@bs]) l

let foldRightMaxStack = 1000


let rec tailLoop f  acc = function
  | [] -> acc
  | h :: t -> tailLoop f (f h acc [@bs]) t

let foldRight f l init =
  let rec loop n = function
    | [] -> init
    | h :: t ->
      if n < foldRightMaxStack then
        f h (loop (n+1) t) [@bs]
      else
        f h (tailLoop f init (rev t)) [@bs]
  in
  loop 0 l


let rec flattenAux  acc  lx =
  match lx with
  | [] -> rev acc
  | y::ys -> flattenAux  (revAppend  y   acc )  ys

let flatten  lx =
  flattenAux  [] lx


let rec filterRevAux f acc xs =
  match xs with
  | [] ->  acc
  | y :: ys ->
    begin match f y [@bs] with
      | false ->   filterRevAux f acc ys
      | true -> filterRevAux f  (y::acc) ys
    end

let filter f xs  =  rev (filterRevAux f [] xs )

let rec filterMapRevAux (f:  'a -> 'b option [@bs]) acc xs =
  match xs with
  | [] ->  acc
  | y :: ys ->
    begin match f y [@bs] with
      | None ->   filterMapRevAux f acc ys
      | Some z -> filterMapRevAux f  (z::acc) ys
    end

let filterMap f xs =
  rev (filterMapRevAux f [] xs)


let rec countByAux f acc xs =
  match xs with
  | [] -> acc
  | y::ys ->
    countByAux f (if f y [@bs] then acc + 1 else acc) ys

let countBy f xs = countByAux f 0 xs

let init n f =
  Js_vector.toList (Js_vector.init n f )

external createUnsafe : int -> 'a Js_vector.t =
  "Array" [@@bs.new]

let toVector xs =
  match xs with
  | [] -> [||]
  | l ->
    let a = createUnsafe (length l) in
    let rec fill i = function
      | [] -> a
      | hd::tl -> Js_array2.unsafe_set a i hd; fill (i+1) tl in
    fill 0 l

let rec equal cmp xs ys =
  match xs,ys with
  | [], [] -> true
  | x::xs, y::ys ->
    if cmp x y [@bs] then equal cmp xs ys
    else false
  | _, _ -> false
